test/functional/dict_notifications_spec.lua

```diff
@@ -0,0 +1,222 @@
+        update('= "xtend"')
+        verify_value({new = 'xtend'})
+        nvim('command', [[
+          call extend(]]..dict_expr..[[, {'watched': 'xtend2', 'watched2': 5, 'watched3': 'a'})
+        ]])
+        verify_value({old = 'xtend', new = 'xtend2'})
+        verify_value({new = 5}, 'watched2')
+        update('')
+        verify_value({old = 'xtend2'})
+        update('', 'watched2')
+        verify_value({old = 5}, 'watched2')
+        update('', 'watched3')
+        verify_echo()
+      end)
+
+      it('is triggered with key patterns', function()
+        source([[
+        call watcheradd(]]..dict_expr..[[, "wat*", "g:Changed")
+        ]])
+        update('= 1')
+        verify_value({new = 1})
+        verify_value({new = 1})
+        update('= 3', 'watched2')
+        verify_value({new = 3}, 'watched2')
+        verify_value({new = 3}, 'watched2')
+        verify_echo()
+        source([[
+        call watcherdel(]]..dict_expr..[[, "wat*", "g:Changed")
+        ]])
+      end)
+
+      -- test a sequence of updates of different types to ensure proper memory
+      -- management(with ASAN)
+      local function test_updates(tests)
+        it('test change sequence', function()
+          local input, output
+          for i = 1, #tests do
+            input, output = unpack(tests[i])
+            update(input)
+            verify_value(output)
+          end
+        end)
+      end
+
+      test_updates({
+        {'= 3', {new = 3}},
+        {'= 6', {old = 3, new = 6}},
+        {'+= 3', {old = 6, new = 9}},
+        {'', {old = 9}}
+      })
+
+      test_updates({
+        {'= "str"', {new = 'str'}},
+        {'= "str2"', {old = 'str', new = 'str2'}},
+        {'.= "2str"', {old = 'str2', new = 'str22str'}},
+        {'', {old = 'str22str'}}
+      })
+
+      test_updates({
+        {'= [1, 2]', {new = {1, 2}}},
+        {'= [1, 2, 3]', {old = {1, 2}, new = {1, 2, 3}}},
+        -- the += will update the list in place, so old and new are the same
+        {'+= [4, 5]', {old = {1, 2, 3, 4, 5}, new = {1, 2, 3, 4, 5}}},
+        {'', {old = {1, 2, 3, 4 ,5}}}
+      })
+
+      test_updates({
+        {'= {"k": "v"}', {new = {k = 'v'}}},
+        {'= {"k1": 2}', {old = {k = 'v'}, new = {k1 = 2}}},
+        {'', {old = {k1 = 2}}},
+      })
+    end)
+  end
+
+  gentests('g:')
+  gentests('b:')
+  gentests('w:')
+  gentests('t:')
+  gentests('g:dict_var', '.', 'let g:dict_var = {}')
+
+  describe('multiple watchers on the same dict/key', function()
+    setup(function()
+      source([[
+      function! g:Watcher1(key, value)
+        call rpcnotify(g:channel, '1', a:key, a:value)
+      endfunction
+      function! g:Watcher2(key, value)
+        call rpcnotify(g:channel, '2', a:key, a:value)
+      endfunction
+      call watcheradd(g:, "key", "g:Watcher1")
+      call watcheradd(g:, "key", "g:Watcher2")
+      ]])
+    end)
+
+    it('invokes all callbacks when the key is changed', function()
```

Comment:
“multiple watchers on the same dict/key” are plural. Also emits “dictionary change notifications multiple watchers on the same dict/key invokes all callbacks when the key is changed”, which is ugly as well.